python/ray/experimental/channel.py

```diff
@@ -187,3 +188,155 @@ def close(self) -> None:
         logger.debug(f"Setting error bit on channel: {self._base_ref}")
         self._ensure_registered_as_writer()
         self._worker.core_worker.experimental_channel_set_error(self._base_ref)
+
+
+# Interfaces for channel I/O.
+class InputReader:
+    def __init__(self, input_channels: List[Channel]):
+        if isinstance(input_channels, List):
+            for chan in input_channels:
+                assert isinstance(chan, Channel)
+            self._has_single_output = False
+        else:
+            assert isinstance(input_channels, Channel)
+            self._has_single_output = True
+            input_channels = [input_channels]
+
+        self._input_channels = input_channels
+        self._closed = False
+        self._num_reads = 0
+
+    def get_num_reads(self) -> int:
+        return self._num_reads
+
+    def start(self):
+        raise NotImplementedError
+
+    def _begin_read_list(self) -> Any:
+        raise NotImplementedError
+
+    def begin_read(self) -> Any:
+        outputs = self._begin_read_list()
+        self._num_reads += 1
+        if self._has_single_output:
+            return outputs[0]
+        else:
+            return outputs
+
+    def end_read(self) -> Any:
+        raise NotImplementedError
+
+    def close(self) -> None:
+        self._closed = True
+        for channel in self._input_channels:
+            channel.close()
+
+
+class SynchronousInputReader(InputReader):
+    def __init__(self, input_channels: List[Channel]):
+        super().__init__(input_channels)
+
+    def start(self):
+        pass
+
+    def _begin_read_list(self) -> Any:
+        return [c.begin_read() for c in self._input_channels]
+
+    def end_read(self) -> Any:
+        for c in self._input_channels:
+            c.end_read()
+
+
+class AwaitableBackgroundOutputReader(InputReader):
```

Comment:
it is confusing to have "Output"Reader is a subclass of "Input"Reader. 

Do we have to associate these classes to output/input? Why don't we just call it writer / reader?

Reply:
Good point! :D